consul/cache.go

```diff
@@ -5,294 +5,128 @@ import (
-					m.cacheServicesAdd(-len(sc.services))
-				}
-			}
 
-		case key := <-next:
-			if _, alreadyPrefetching := prefetches[key]; !alreadyPrefetching {
-				prefetches[key] = c.spawn(key, ready, done, next)
-				m := key.metrics()
-				m.cachePrefetchesInc()
-			}
-
-		case sc := <-ready:
-			if prefetches[sc.key] != sc {
-				sc.close()
-			} else {
-				delete(prefetches, sc.key)
-				expired := caches[sc.key]
-				caches[sc.key] = sc
-				m := sc.key.metrics()
-
-				if expired != nil {
-					if expired.negative() {
-						m.cacheSizeAddDenial(-1)
-					} else {
-						m.cacheSizeAddSuccess(-1)
-						m.cacheServicesAdd(-len(expired.services))
-					}
-					expired.close()
-				}
-
-				if sc.negative() {
-					m.cacheSizeAddDenial(1)
-				} else {
-					m.cacheSizeAddSuccess(1)
-					m.cacheServicesAdd(len(sc.services))
-					m.cacheFetchSizesObserve(len(sc.services))
-				}
-			}
-
-		case r, ok := <-reqs:
-			if !ok {
-				break dispatchRequests
-			}
+	mutex    sync.RWMutex
+	entries  map[key]*entry
+	lookups  atomicIndex
+	cleanups atomicLock
+}
 
-			sc, hit := caches[r.key]
-			if !hit {
-				sc, hit = prefetches[r.key]
-				if !hit {
-					sc = c.spawn(r.key, ready, done, next)
-					prefetches[r.key] = sc
-					m := r.key.metrics()
-					m.cacheMissesInc()
-				}
-			}
+func (c *cache) prefetchDeadlineOf(e *entry) time.Time {
+	d := float64(c.prefetchDuration) * (float64(c.prefetchPercentage) / 1000)
+	return e.exp.Add(-time.Duration(d))
+}
 
-			select {
-			case sc.reqs <- r:
-			default:
-				r.reject(errTooManyRequests)
+func (c *cache) lookup(ctx context.Context, k key, now time.Time) (srv service, ttl time.Duration, err error) {
+	e := c.grab(k, now)
+	i := e.index.incr() - 1
+
+	// Note: the implementation of this check should be changed to take prefetchAmount
+	// into account, but it requires maintaining more state to implement it right, which
+	// is not immediately needed since consul service names looked up in production are
+	// all very popular.
+	if i == 0 || (i >= uint32(c.prefetchAmount)) && now.After(c.prefetchDeadlineOf(e)) {
+		if e.lock.tryLock() {
+			srv, err := c.load(k)
+			e.lock.unlock()
+
+			if e.once.tryLock() {
+				e.srv = srv
+				e.err = err
+				close(e.ready)
+			} else if err == nil {
+				c.update(k, &entry{
+					srv:   srv,
+					exp:   now.Add(c.ttl),
+					ready: e.ready, // already closed
+					index: 1,       // can't be zero to avoid refetching on next lookup
+				})
 			}
 		}
 	}
 
-	for _, sc := range prefetches {
-		sc.close()
-	}
-
-	for _, sc := range caches {
-		sc.close()
-	}
-
-	for len(caches) != 0 || len(prefetches) != 0 {
-		select {
-		case <-next:
-			// ignore
-
-		case sc := <-done:
-			if caches[sc.key] == sc {
-				delete(caches, sc.key)
-			}
-
-		case sc := <-ready:
-			if prefetches[sc.key] == sc {
-				delete(prefetches, sc.key)
-			}
+	// Every 1000 lookups the cache removes expired entries.
+	if (c.lookups.incr() % 1000) == 0 {
+		if c.cleanups.tryLock() {
```

Comment:
this seems to make it unnecessarily difficult to understand what will happen - are we really so worried about lock contention here, given we are only locking once in 1000 lookups times? Why not a regular mutex?